primitives/core/src/paired_crypto.rs

```diff
@@ -0,0 +1,124 @@
+	pub const CRYPTO_ID: CryptoTypeId = CryptoTypeId(*b"ecb7");
+
+	/// BLS12-377 key pair.
+	#[cfg(feature = "full_crypto")]
+	//pub type Pair = super::Pair<ecdsa:Pair, bls377:Pair>;
+	/// BLS12-377 public key.
+	pub type Public = super::Public<ecdsa::Public, ed25519::Public, 32, 32>;
+	// /// BLS12-377 signature.
+	//pub type Signature = super::Signature<ecdsa:Signature, bls377:Signature>;
+
+	// impl super::HardJunctionId for TinyBLS377 {
+	// 	const ID: &'static str = "BLS12377HDKD";
+	// }
+}
+
+/// A secret seed.
+///
+/// It's not called a "secret key" because ring doesn't expose the secret keys
+/// of the key pair (yeah, dumb); as such we're forced to remember the seed manually if we
+/// will need it later (such as for HDKD).
+#[cfg(feature = "full_crypto")]
+//type Seed = [u8; SECRET_KEY_SERIALIZED_SIZE];
+
+pub trait PublicKeyBound: TraitPublic + sp_std::hash::Hash + ByteArray  {}
+/// A public key.
+#[derive(Clone, Encode, Decode, MaxEncodedLen, TypeInfo, PartialEq, Eq, PartialOrd, Ord)]
+#[scale_info(skip_type_params(T))]
+pub struct Public<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_LEN: usize, const RIGHT_LEN: usize,> (LeftPublic, RightPublic);
+
+#[cfg(feature = "full_crypto")]
+impl<LeftPublic: PublicKeyBound + UncheckedFrom<[u8; LEFT_LEN]>, RightPublic: PublicKeyBound + UncheckedFrom<[u8; RIGHT_LEN]>, const LEFT_LEN: usize, const RIGHT_LEN: usize> sp_std::hash::Hash for Public<LeftPublic, RightPublic, LEFT_LEN, RIGHT_LEN> {
+  	fn hash<H: sp_std::hash::Hasher>(&self, state: &mut H) {
+ 	    self.0.hash(state);
+	    self.1.hash(state);
+	}
+}
+
+impl<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_LEN: usize, const RIGHT_LEN: usize> ByteArray  for Public<LeftPublic, RightPublic, LEFT_LEN, RIGHT_LEN> where for<'a>  Public<LeftPublic, RightPublic, LEFT_LEN, RIGHT_LEN>: TryFrom<&'a [u8], Error = ()> + AsRef<[u8]> + AsMut<[u8]>  {
+     const LEN: usize = LeftPublic::LEN + RightPublic::LEN;
+}
+
+impl<'a,LeftPublic: PublicKeyBound + UncheckedFrom<[u8; LEFT_LEN]>, RightPublic: PublicKeyBound + UncheckedFrom<[u8; RIGHT_LEN]>, const LEFT_LEN: usize, const RIGHT_LEN: usize>  TryFrom<&'a[u8]> for Public<LeftPublic, RightPublic, LEFT_LEN, RIGHT_LEN> {
+    type Error = ();
+
+    fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+ 		if data.len() != 	LEFT_LEN + RIGHT_LEN {
+			return Err(())
+		}
+	
+ 		let mut r0 = [0u8;  LEFT_LEN];
+ 		let mut r1 = [0u8; RIGHT_LEN];
+		r0.copy_from_slice(&data[0..LEFT_LEN]);  
+ 		r1.copy_from_slice(&data[LEFT_LEN..RIGHT_LEN]);
+		Ok(Self(LeftPublic::unchecked_from(r0),RightPublic::unchecked_from(r1)))
+    }
+}
+
+// impl<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_LEN: usize, const RIGHT_LEN: usize> AsMut<[u8]> for Public<LeftPublic, RightPublic, LEFT_LEN, RIGHT_LEN> {
+//  	fn as_mut(&mut self) -> &mut [u8] {
+//  	 &mut [self.0.as_mut(), self.1.as_mut()].concat()
+//  	}
+// }
```

Comment:
@davxy it seems that the second representation `(LeftPublic: PublicBound, RightPublic: PublicBound)` does not work because we can not implement `AsRef<u8>` and `AsMut<u8>` as we don't have a consecutive representation of the sub-elements as [these people suggests](https://users.rust-lang.org/t/how-to-implement-the-trait-std-asref-u8-for-generic-structure/33554). Should I go back to the first solution:
```
pub struct PublicWithInner<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_PLUS_RIGHT_SIZE: usize> {
    inner: [u8; LEFT_PLUS_RIGHT_SIZE],
 	_phantom: PhantomData<fn() -> (LeftPublic, RightPublic)>,
}
```

or should I keep both representations at the same time:
```
pub struct Public<LeftPublic: PublicKeyBound, RightPublic: PublicKeyBound, const LEFT_PLUS_RIGHT_SIZE: usize> {
    left_public : LeftPublic,
    right_public: RightPublic,
    inner: [u8; LEFT_PLUS_RIGHT_SIZE],

}
```
or you have a solution to building non-ephemeral slice in a function that works here?

🙏